MAINT,API: ignore and NULL fasttake/fastputmask ArrFuncs slots

seberg opened the pull request:
Always NULL the fasttake slot, and instead move the optimized versions
by doing itemsize specialization in the Take function itself.

---

I still need to check that there is now speed regression.

seberg committed:
MAINT,API: Do not use fasttake slot of ArrFuncs and null it

Always NULL the fasttake slot, and instead move the optimized versions
by doing itemsize specialization in the Take function itself.

eric-wieser commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -133,107 +244,46 @@ PyArray_TakeFrom(PyArrayObject *self0, PyObject *indices0, int axis,
-        NPY_END_THREADS;
+    int res;
+    /* Specifically list all builtin itemsizes to allow compiler optimization */
+    if (!needs_refcounting && (itemsize == 1)) {
```
Can you add an outer `if(!needs_refcounting)` here to make  all these conditions shorter?

seberg replied:
Yeah, have to duplicate the "default" branch than maybe, but that would be fine I guess.

seberg replied:
Made it a switch statement, but right now it seems like the optimization is not kicking in (or at least not to full effect), hmmmm.

seberg replied:
nvm, the single element case was only there in the fastclip definitions, that is probably missing.

eric-wieser commented on numpy/core/src/multiarray/arraytypes.h:
```diff
@@ -28,4 +28,31 @@ small_correlate(const char * d_, npy_intp dstride,
+NPY_NO_EXPORT int
+npy_fasttake_1(
+        char *, char *, npy_intp *, npy_intp, npy_intp, npy_intp, npy_intp,
+        NPY_CLIPMODE);
```
Where's the definition of this?

seberg replied:
Opst, that is nonsense I forgot to delete again.

eric-wieser commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -133,107 +243,50 @@ PyArray_TakeFrom(PyArrayObject *self0, PyObject *indices0, int axis,
+            res = npy_fasttake_impl(
+                    dest, src, indices_data, n, m, max_item, nelem, chunk,
+                    clipmode, itemsize, needs_refcounting, dtype, axis);
     }
```
Can you move this entire function to `npy_fasttake`, which can have a very clear docstring of something like:
```
/* npy_fasttake_impl, instantiated repeatedly in different branches for optimization reasons */
```

eric-wieser commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -21,11 +21,122 @@
+        NPY_CLIPMODE clipmode, npy_intp itemsize, int needs_refcounting,
+        PyArray_Descr *dtype, int axis)
+{
+    npy_intp i, j, k;
```
Since we're touching and likely not backporting, can we push these definitions into their respective loops?

seberg commented:
Think I addressed those, will squash in a bit. Maybe it is due to debug build, but now changed it to split up on chunk (because why not, it just speeds up certain cases). And it seems `memmove` is much faster even for 3 bools (which miss the specialized loops), due to `memmove` rather than a loop for 0 to 3. So overall, it seems like this is a pretty big speed win actually, plus it makes the error message cleaner...

seberg commented:
I want to do the same thing for `fastput`, planning to do a new PR for that though right now.

seberg commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -21,11 +21,171 @@
+                    dest, src, indices, n, m, max_item, nelem, 1,
+                    clipmode, itemsize, 0, dtype, axis);
+        }
+        else if (chunk == 2) {
```
(I will remove the else when squashing)

seberg committed:
MAINT,API: Do not use fastputmask dtype slot and always NULL it

This removes the use of the fastputmask ArrFuncs slot from within
NumPy and always leaves it NULL.

eric-wieser commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -21,11 +21,171 @@
+            return npy_fasttake_impl(
+                    dest, src, indices, n, m, max_item, nelem, 32,
+                    clipmode, itemsize, 0, dtype, axis);
+        }
```
it's tempting to leave `chunk` and `needs_refcounting` inline here, to make all of these the same.

seberg replied:
Right, had removed that due to the switch statement not being simple enough for the compiler, but with the if it doesn't matter.

eric-wieser commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -21,11 +21,171 @@
+        PyArray_Descr *dtype, int axis)
+{
+    char *tmp_src;
+    npy_intp tmp;
```
nit: could move these into their loops too and save three lines

seberg commented:
The current timings are below. Basically, take is slower for 3 elements (which is too bad, since it was a strength of the old take compared to advanced indexing. That could probably be resolved by putting in the loop over the inner-size instead of the large chunk memmove.

For some reason complex256 is optimized much better in take (maybe alignment? maybe just compiler fluctuations). `fastputmask` is a bit slower, not sure why, but usually it is not a huge amount. And it is not a very common function.

<details>

<pre>       before           after         ratio
     [01123f61]       [6ec673a3]
<font color="#CC0000">+     6.32±0.05μs       10.9±0.2μs     1.72  </font>bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;clip&apos;, &apos;complex256&apos;)
<font color="#CC0000">+     6.29±0.06μs      10.7±0.05μs     1.70  </font>bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;wrap&apos;, &apos;complex256&apos;)
<font color="#CC0000">+     3.78±0.07μs      6.22±0.04μs     1.64  </font>bench_itemselection.Take.time_contiguous((1000, 1), &apos;wrap&apos;, &apos;complex256&apos;)
<font color="#CC0000">+     3.59±0.06μs      5.86±0.06μs     1.63  </font>bench_itemselection.PutMask.time_dense(False, &apos;complex256&apos;)
<font color="#CC0000">+      6.78±0.1μs       10.8±0.1μs     1.59  </font>bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;raise&apos;, &apos;complex256&apos;)
<font color="#CC0000">+      3.91±0.1μs      6.15±0.07μs     1.57  </font>bench_itemselection.Take.time_contiguous((1000, 1), &apos;clip&apos;, &apos;complex256&apos;)
<font color="#CC0000">+     4.07±0.07μs      6.27±0.03μs     1.54  </font>bench_itemselection.Take.time_contiguous((1000, 1), &apos;raise&apos;, &apos;complex256&apos;)
<font color="#CC0000">+     6.11±0.07μs      8.91±0.04μs     1.46  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;wrap&apos;, &apos;complex256&apos;)
<font color="#CC0000">+      8.85±0.3μs       12.6±0.3μs     1.42  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;wrap&apos;, &apos;complex256&apos;)
<font color="#CC0000">+      8.54±0.1μs      12.1±0.08μs     1.42  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;raise&apos;, &apos;complex256&apos;)
<font color="#CC0000">+      8.48±0.1μs       12.0±0.2μs     1.42  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;clip&apos;, &apos;complex256&apos;)
<font color="#CC0000">+     6.18±0.08μs      8.38±0.07μs     1.35  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;clip&apos;, &apos;complex256&apos;)
<font color="#CC0000">+      5.94±0.1μs      8.01±0.04μs     1.35  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;clip&apos;, &apos;int16&apos;)
<font color="#CC0000">+      6.35±0.1μs       8.49±0.1μs     1.34  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;raise&apos;, &apos;complex256&apos;)
         6.26±1μs       8.12±0.2μs    ~1.30  bench_itemselection.Take.time_contiguous((1000, 3), &apos;raise&apos;, &apos;int16&apos;)
<font color="#CC0000">+      6.28±0.1μs       7.92±0.1μs     1.26  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;wrap&apos;, &apos;complex128&apos;)
<font color="#CC0000">+     5.92±0.07μs       7.34±0.2μs     1.24  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;raise&apos;, &apos;complex64&apos;)
       6.11±0.1μs      7.53±0.03μs    ~1.23  bench_itemselection.Take.time_contiguous((1000, 3), &apos;wrap&apos;, &apos;float16&apos;)
<font color="#CC0000">+     5.98±0.05μs       7.19±0.1μs     1.20  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;clip&apos;, &apos;complex64&apos;)
<font color="#CC0000">+     5.99±0.05μs      7.19±0.09μs     1.20  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;wrap&apos;, &apos;float64&apos;)
<font color="#CC0000">+     5.27±0.05μs      6.29±0.06μs     1.19  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;wrap&apos;, &apos;complex128&apos;)
<font color="#CC0000">+      5.97±0.1μs      7.03±0.03μs     1.18  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;raise&apos;, &apos;int32&apos;)
<font color="#CC0000">+     5.96±0.03μs       7.02±0.1μs     1.18  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;wrap&apos;, &apos;float32&apos;)
<font color="#CC0000">+     6.03±0.02μs      7.05±0.04μs     1.17  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;clip&apos;, &apos;int32&apos;)
<font color="#CC0000">+     6.19±0.09μs       7.23±0.4μs     1.17  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;raise&apos;, &apos;int64&apos;)
<font color="#CC0000">+     6.03±0.05μs      6.99±0.04μs     1.16  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;wrap&apos;, &apos;int32&apos;)
<font color="#CC0000">+     6.32±0.05μs       7.25±0.2μs     1.15  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;wrap&apos;, &apos;complex64&apos;)
<font color="#CC0000">+     2.04±0.01μs      2.31±0.01μs     1.13  </font>bench_itemselection.PutMask.time_dense(True, &apos;float16&apos;)
<font color="#CC0000">+      7.10±0.2μs      8.02±0.03μs     1.13  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;raise&apos;, &apos;float16&apos;)
<font color="#CC0000">+     2.06±0.01μs      2.32±0.01μs     1.13  </font>bench_itemselection.PutMask.time_dense(True, &apos;float32&apos;)
<font color="#CC0000">+      7.11±0.3μs       8.02±0.3μs     1.13  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;clip&apos;, &apos;float16&apos;)
<font color="#CC0000">+     2.66±0.02μs         2.98±0μs     1.12  </font>bench_itemselection.PutMask.time_dense(False, &apos;int64&apos;)
<font color="#CC0000">+     2.64±0.01μs      2.95±0.02μs     1.12  </font>bench_itemselection.PutMask.time_dense(False, &apos;complex64&apos;)
<font color="#CC0000">+     2.56±0.01μs      2.86±0.03μs     1.12  </font>bench_itemselection.PutMask.time_dense(False, &apos;float16&apos;)
<font color="#CC0000">+     2.09±0.01μs      2.34±0.01μs     1.12  </font>bench_itemselection.PutMask.time_dense(True, &apos;int64&apos;)
<font color="#CC0000">+     2.08±0.04μs      2.30±0.01μs     1.11  </font>bench_itemselection.PutMask.time_dense(True, &apos;int16&apos;)
<font color="#CC0000">+     2.58±0.01μs      2.85±0.01μs     1.11  </font>bench_itemselection.PutMask.time_dense(False, &apos;int32&apos;)
<font color="#CC0000">+     2.67±0.01μs      2.94±0.02μs     1.10  </font>bench_itemselection.PutMask.time_dense(False, &apos;float64&apos;)
      2.57±0.01μs      2.82±0.01μs     1.10  bench_itemselection.PutMask.time_dense(False, &apos;int16&apos;)
      2.59±0.01μs      2.84±0.01μs     1.10  bench_itemselection.PutMask.time_dense(False, &apos;float32&apos;)
      4.91±0.03μs      5.30±0.02μs     1.08  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;clip&apos;, &apos;complex128&apos;)
      3.23±0.02μs      3.48±0.04μs     1.08  bench_itemselection.Take.time_contiguous((1000, 1), &apos;clip&apos;, &apos;complex128&apos;)
      7.25±0.02μs       7.71±0.1μs     1.06  bench_itemselection.Take.time_contiguous((1000, 3), &apos;raise&apos;, &apos;complex128&apos;)
      2.31±0.01μs      2.40±0.01μs     1.04  bench_itemselection.PutMask.time_sparse(False, &apos;complex256&apos;)
      7.33±0.06μs      7.52±0.02μs     1.03  bench_itemselection.Take.time_contiguous((1000, 3), &apos;wrap&apos;, &apos;int16&apos;)
      4.42±0.06μs      4.51±0.07μs     1.02  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;clip&apos;, &apos;float64&apos;)
      2.61±0.03μs      2.66±0.01μs     1.02  bench_itemselection.PutMask.time_dense(True, &apos;complex128&apos;)
      7.20±0.09μs       7.32±0.1μs     1.02  bench_itemselection.Take.time_contiguous((1000, 3), &apos;raise&apos;, &apos;float64&apos;)
      4.92±0.01μs      4.99±0.09μs     1.02  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;raise&apos;, &apos;float64&apos;)
      3.21±0.02μs      3.25±0.02μs     1.01  bench_itemselection.Take.time_contiguous((1000, 1), &apos;raise&apos;, &apos;int16&apos;)
      2.30±0.01μs      2.34±0.01μs     1.01  bench_itemselection.PutMask.time_sparse(False, &apos;float64&apos;)
      4.43±0.03μs       4.48±0.1μs     1.01  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;clip&apos;, &apos;complex64&apos;)
      3.29±0.02μs      3.32±0.01μs     1.01  bench_itemselection.Take.time_contiguous((1000, 1), &apos;raise&apos;, &apos;float64&apos;)
      2.29±0.01μs      2.31±0.01μs     1.01  bench_itemselection.PutMask.time_dense(True, &apos;int32&apos;)
      4.44±0.01μs      4.47±0.09μs     1.01  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;clip&apos;, &apos;int64&apos;)
      7.23±0.01μs      7.28±0.09μs     1.01  bench_itemselection.Take.time_contiguous((1000, 3), &apos;clip&apos;, &apos;complex128&apos;)
      3.21±0.02μs      3.23±0.01μs     1.01  bench_itemselection.Take.time_contiguous((1000, 1), &apos;raise&apos;, &apos;float16&apos;)
      2.94±0.03μs      2.96±0.02μs     1.01  bench_itemselection.Take.time_contiguous((1000, 1), &apos;clip&apos;, &apos;int16&apos;)
      2.32±0.01μs      2.33±0.01μs     1.01  bench_itemselection.PutMask.time_dense(True, &apos;float64&apos;)
      2.32±0.01μs      2.33±0.01μs     1.00  bench_itemselection.PutMask.time_dense(True, &apos;complex64&apos;)
      3.03±0.04μs      3.04±0.06μs     1.00  bench_itemselection.Take.time_contiguous((1000, 1), &apos;clip&apos;, &apos;float32&apos;)
      3.31±0.02μs      3.32±0.01μs     1.00  bench_itemselection.Take.time_contiguous((1000, 1), &apos;raise&apos;, &apos;int64&apos;)
      4.33±0.02μs      4.33±0.02μs     1.00  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;clip&apos;, &apos;float16&apos;)
      4.30±0.03μs      4.31±0.02μs     1.00  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;clip&apos;, &apos;int16&apos;)
      3.28±0.03μs      3.28±0.02μs     1.00  bench_itemselection.Take.time_contiguous((1000, 1), &apos;raise&apos;, &apos;int32&apos;)
      3.05±0.02μs      3.05±0.02μs     1.00  bench_itemselection.Take.time_contiguous((1000, 1), &apos;clip&apos;, &apos;float64&apos;)
      4.88±0.02μs      4.88±0.02μs     1.00  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;raise&apos;, &apos;float32&apos;)
      4.81±0.02μs      4.81±0.03μs     1.00  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;raise&apos;, &apos;int16&apos;)
      4.88±0.02μs      4.87±0.02μs     1.00  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;raise&apos;, &apos;int32&apos;)
      3.34±0.03μs      3.33±0.02μs     1.00  bench_itemselection.Take.time_contiguous((1000, 1), &apos;raise&apos;, &apos;complex64&apos;)
      4.98±0.01μs      4.96±0.03μs     1.00  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;raise&apos;, &apos;complex64&apos;)
      4.36±0.02μs      4.34±0.03μs     1.00  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;clip&apos;, &apos;float32&apos;)
      2.96±0.03μs      2.95±0.03μs     1.00  bench_itemselection.Take.time_contiguous((1000, 1), &apos;clip&apos;, &apos;float16&apos;)
      4.84±0.04μs      4.81±0.02μs     1.00  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;raise&apos;, &apos;float16&apos;)
      2.32±0.01μs      2.30±0.02μs     0.99  bench_itemselection.PutMask.time_sparse(True, &apos;int16&apos;)
         2.33±0μs      2.32±0.01μs     0.99  bench_itemselection.PutMask.time_sparse(False, &apos;int16&apos;)
      3.28±0.02μs      3.26±0.03μs     0.99  bench_itemselection.Take.time_contiguous((1000, 1), &apos;raise&apos;, &apos;float32&apos;)
      3.07±0.02μs      3.05±0.02μs     0.99  bench_itemselection.Take.time_contiguous((1000, 1), &apos;clip&apos;, &apos;complex64&apos;)
      4.96±0.03μs      4.92±0.02μs     0.99  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;raise&apos;, &apos;int64&apos;)
      3.09±0.02μs      3.06±0.01μs     0.99  bench_itemselection.Take.time_contiguous((1000, 1), &apos;clip&apos;, &apos;int64&apos;)
      7.17±0.08μs       7.09±0.1μs     0.99  bench_itemselection.Take.time_contiguous((1000, 3), &apos;raise&apos;, &apos;float32&apos;)
      3.28±0.03μs      3.21±0.01μs     0.98  bench_itemselection.PutMask.time_dense(False, &apos;complex128&apos;)
      3.14±0.02μs      3.02±0.04μs     0.96  bench_itemselection.Take.time_contiguous((1000, 1), &apos;clip&apos;, &apos;int32&apos;)
      3.52±0.02μs      3.39±0.02μs     0.96  bench_itemselection.Take.time_contiguous((1000, 1), &apos;raise&apos;, &apos;complex128&apos;)
      7.40±0.08μs       7.08±0.1μs     0.96  bench_itemselection.Take.time_contiguous((1000, 3), &apos;clip&apos;, &apos;float32&apos;)
      7.49±0.03μs       7.16±0.2μs     0.96  bench_itemselection.Take.time_contiguous((1000, 3), &apos;clip&apos;, &apos;float64&apos;)
       7.59±0.1μs      7.21±0.04μs     0.95  bench_itemselection.Take.time_contiguous((1000, 3), &apos;clip&apos;, &apos;int64&apos;)
       6.59±0.2μs      6.23±0.01μs     0.95  bench_itemselection.Take.time_contiguous((1000, 2), &apos;clip&apos;, &apos;complex128&apos;)
      4.60±0.02μs      4.32±0.02μs     0.94  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;clip&apos;, &apos;int32&apos;)
      2.89±0.01μs      2.69±0.01μs     0.93  bench_itemselection.Take.time_contiguous((1000, 1), &apos;wrap&apos;, &apos;int16&apos;)
      3.35±0.01μs      3.11±0.05μs     0.93  bench_itemselection.Take.time_contiguous((1000, 1), &apos;wrap&apos;, &apos;complex128&apos;)
      5.46±0.04μs      5.06±0.07μs     0.93  bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;raise&apos;, &apos;complex128&apos;)
      6.78±0.05μs      6.24±0.08μs     0.92  bench_itemselection.Take.time_contiguous((1000, 2), &apos;raise&apos;, &apos;complex128&apos;)
      2.91±0.02μs      2.68±0.01μs     0.92  bench_itemselection.Take.time_contiguous((1000, 1), &apos;wrap&apos;, &apos;float16&apos;)
<font color="#4E9A06">-     2.98±0.04μs      2.71±0.02μs     0.91  </font>bench_itemselection.Take.time_contiguous((1000, 1), &apos;wrap&apos;, &apos;float32&apos;)
<font color="#4E9A06">-     8.06±0.06μs      7.31±0.06μs     0.91  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;wrap&apos;, &apos;int64&apos;)
<font color="#4E9A06">-     3.03±0.01μs      2.75±0.02μs     0.91  </font>bench_itemselection.Take.time_contiguous((1000, 1), &apos;wrap&apos;, &apos;float64&apos;)
<font color="#4E9A06">-     3.03±0.03μs      2.73±0.02μs     0.90  </font>bench_itemselection.Take.time_contiguous((1000, 1), &apos;wrap&apos;, &apos;int32&apos;)
<font color="#4E9A06">-     4.27±0.01μs      3.82±0.01μs     0.90  </font>bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;wrap&apos;, &apos;int16&apos;)
<font color="#4E9A06">-     4.42±0.01μs      3.94±0.02μs     0.89  </font>bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;wrap&apos;, &apos;float64&apos;)
<font color="#4E9A06">-        4.28±0μs      3.82±0.02μs     0.89  </font>bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;wrap&apos;, &apos;float16&apos;)
<font color="#4E9A06">-     3.10±0.02μs      2.76±0.01μs     0.89  </font>bench_itemselection.Take.time_contiguous((1000, 1), &apos;wrap&apos;, &apos;int64&apos;)
<font color="#4E9A06">-     2.29±0.01μs      2.03±0.01μs     0.89  </font>bench_itemselection.PutMask.time_sparse(True, &apos;complex128&apos;)
<font color="#4E9A06">-     4.34±0.02μs      3.85±0.01μs     0.89  </font>bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;wrap&apos;, &apos;int32&apos;)
<font color="#4E9A06">-     2.31±0.02μs      2.04±0.01μs     0.88  </font>bench_itemselection.PutMask.time_sparse(True, &apos;complex64&apos;)
<font color="#4E9A06">-     2.31±0.01μs         2.04±0μs     0.88  </font>bench_itemselection.PutMask.time_sparse(True, &apos;float64&apos;)
<font color="#4E9A06">-     3.16±0.02μs      2.78±0.02μs     0.88  </font>bench_itemselection.Take.time_contiguous((1000, 1), &apos;wrap&apos;, &apos;complex64&apos;)
<font color="#4E9A06">-     5.03±0.01μs      4.43±0.05μs     0.88  </font>bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;wrap&apos;, &apos;complex128&apos;)
<font color="#4E9A06">-     2.31±0.02μs      2.03±0.01μs     0.88  </font>bench_itemselection.PutMask.time_sparse(True, &apos;int32&apos;)
<font color="#4E9A06">-     4.45±0.04μs      3.90±0.01μs     0.88  </font>bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;wrap&apos;, &apos;int64&apos;)
<font color="#4E9A06">-     4.37±0.04μs      3.83±0.02μs     0.88  </font>bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;wrap&apos;, &apos;float32&apos;)
<font color="#4E9A06">-     4.46±0.03μs      3.90±0.04μs     0.87  </font>bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;wrap&apos;, &apos;complex64&apos;)
<font color="#4E9A06">-     2.34±0.01μs      2.03±0.01μs     0.87  </font>bench_itemselection.PutMask.time_sparse(True, &apos;longfloat&apos;)
<font color="#4E9A06">-     2.82±0.02μs         2.31±0μs     0.82  </font>bench_itemselection.PutMask.time_sparse(False, &apos;int32&apos;)
<font color="#4E9A06">-     2.98±0.01μs         2.31±0μs     0.78  </font>bench_itemselection.PutMask.time_sparse(False, &apos;longfloat&apos;)
<font color="#4E9A06">-     3.02±0.02μs      2.34±0.01μs     0.77  </font>bench_itemselection.PutMask.time_sparse(False, &apos;complex128&apos;)
<font color="#4E9A06">-     3.01±0.01μs      2.33±0.01μs     0.77  </font>bench_itemselection.PutMask.time_sparse(False, &apos;int64&apos;)
<font color="#4E9A06">-     3.03±0.05μs         2.33±0μs     0.77  </font>bench_itemselection.PutMask.time_sparse(False, &apos;complex64&apos;)
<font color="#4E9A06">-     3.06±0.02μs      2.33±0.02μs     0.76  </font>bench_itemselection.PutMask.time_sparse(False, &apos;float16&apos;)
<font color="#4E9A06">-     3.07±0.01μs      2.32±0.01μs     0.76  </font>bench_itemselection.PutMask.time_sparse(False, &apos;float32&apos;)
<font color="#4E9A06">-     3.04±0.09μs         2.29±0μs     0.75  </font>bench_itemselection.PutMask.time_sparse(True, &apos;float16&apos;)
<font color="#4E9A06">-     7.06±0.03μs      5.13±0.08μs     0.73  </font>bench_itemselection.PutMask.time_dense(True, &apos;complex256&apos;)
<font color="#4E9A06">-     5.02±0.01μs      3.52±0.06μs     0.70  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;clip&apos;, &apos;complex64&apos;)
<font color="#4E9A06">-     9.02±0.03μs      6.30±0.06μs     0.70  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;clip&apos;, &apos;longfloat&apos;)
<font color="#4E9A06">-     3.07±0.03μs      2.12±0.01μs     0.69  </font>bench_itemselection.PutMask.time_sparse(True, &apos;complex256&apos;)
<font color="#4E9A06">-     4.95±0.02μs      3.37±0.03μs     0.68  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;raise&apos;, &apos;float64&apos;)
<font color="#4E9A06">-     9.20±0.02μs      6.23±0.03μs     0.68  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;raise&apos;, &apos;longfloat&apos;)
<font color="#4E9A06">-     3.02±0.05μs      2.03±0.01μs     0.67  </font>bench_itemselection.PutMask.time_sparse(True, &apos;float32&apos;)
<font color="#4E9A06">-     3.04±0.02μs      2.04±0.01μs     0.67  </font>bench_itemselection.PutMask.time_sparse(True, &apos;int64&apos;)
<font color="#4E9A06">-     5.23±0.05μs      3.39±0.08μs     0.65  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;raise&apos;, &apos;int32&apos;)
<font color="#4E9A06">-     11.9±0.07μs       7.71±0.2μs     0.65  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;raise&apos;, &apos;longfloat&apos;)
<font color="#4E9A06">-     5.41±0.04μs      3.49±0.02μs     0.64  </font>bench_itemselection.Take.time_contiguous((1000, 1), &apos;clip&apos;, &apos;longfloat&apos;)
<font color="#4E9A06">-     5.33±0.06μs      3.39±0.01μs     0.64  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;raise&apos;, &apos;int64&apos;)
<font color="#4E9A06">-     5.30±0.06μs      3.37±0.03μs     0.63  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;raise&apos;, &apos;complex64&apos;)
<font color="#4E9A06">-     5.56±0.05μs      3.48±0.01μs     0.62  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;clip&apos;, &apos;float64&apos;)
<font color="#4E9A06">-     5.61±0.02μs      3.46±0.02μs     0.62  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;clip&apos;, &apos;int64&apos;)
<font color="#4E9A06">-     5.35±0.04μs      3.29±0.03μs     0.61  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;raise&apos;, &apos;int16&apos;)
<font color="#4E9A06">-     5.07±0.05μs      3.11±0.03μs     0.61  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;wrap&apos;, &apos;float64&apos;)
<font color="#4E9A06">-     5.00±0.03μs      3.07±0.03μs     0.61  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;wrap&apos;, &apos;complex64&apos;)
<font color="#4E9A06">-     4.96±0.08μs      3.01±0.03μs     0.61  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;clip&apos;, &apos;int16&apos;)
<font color="#4E9A06">-     10.7±0.09μs      6.37±0.03μs     0.60  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;wrap&apos;, &apos;longfloat&apos;)
<font color="#4E9A06">-     4.45±0.01μs      2.65±0.01μs     0.60  </font>bench_itemselection.PutMask.time_dense(True, &apos;longfloat&apos;)
<font color="#4E9A06">-     12.3±0.07μs       7.28±0.2μs     0.59  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;clip&apos;, &apos;longfloat&apos;)
<font color="#4E9A06">-     9.06±0.03μs      5.33±0.06μs     0.59  </font>bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;clip&apos;, &apos;longfloat&apos;)
<font color="#4E9A06">-     5.24±0.08μs      3.08±0.02μs     0.59  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;clip&apos;, &apos;int32&apos;)
<font color="#4E9A06">-     13.7±0.04μs       7.92±0.1μs     0.58  </font>bench_itemselection.Take.time_contiguous((1000, 3), &apos;wrap&apos;, &apos;longfloat&apos;)
<font color="#4E9A06">-     4.94±0.02μs      2.76±0.02μs     0.56  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;wrap&apos;, &apos;float32&apos;)
<font color="#4E9A06">-     4.95±0.03μs      2.75±0.02μs     0.56  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;wrap&apos;, &apos;int32&apos;)
<font color="#4E9A06">-     4.93±0.02μs      2.72±0.03μs     0.55  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;wrap&apos;, &apos;float16&apos;)
<font color="#4E9A06">-     5.85±0.04μs      3.14±0.02μs     0.54  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;wrap&apos;, &apos;int64&apos;)
<font color="#4E9A06">-     5.91±0.05μs      3.08±0.03μs     0.52  </font>bench_itemselection.Take.time_contiguous((1000, 1), &apos;wrap&apos;, &apos;longfloat&apos;)
<font color="#4E9A06">-      6.43±0.2μs      3.29±0.03μs     0.51  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;raise&apos;, &apos;float16&apos;)
<font color="#4E9A06">-     6.84±0.03μs      3.41±0.05μs     0.50  </font>bench_itemselection.Take.time_contiguous((1000, 1), &apos;raise&apos;, &apos;longfloat&apos;)
<font color="#4E9A06">-     6.08±0.06μs      3.03±0.04μs     0.50  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;clip&apos;, &apos;float16&apos;)
<font color="#4E9A06">-     6.52±0.02μs      3.21±0.01μs     0.49  </font>bench_itemselection.PutMask.time_dense(False, &apos;longfloat&apos;)
<font color="#4E9A06">-     6.86±0.03μs      3.30±0.02μs     0.48  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;raise&apos;, &apos;float32&apos;)
<font color="#4E9A06">-      6.04±0.2μs      2.73±0.01μs     0.45  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;wrap&apos;, &apos;int16&apos;)
<font color="#4E9A06">-     9.98±0.05μs      4.48±0.04μs     0.45  </font>bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;wrap&apos;, &apos;longfloat&apos;)
<font color="#4E9A06">-     6.92±0.02μs      3.08±0.02μs     0.45  </font>bench_itemselection.Take.time_contiguous((1000, 2), &apos;clip&apos;, &apos;float32&apos;)
<font color="#4E9A06">-     12.0±0.03μs      5.10±0.04μs     0.43  </font>bench_itemselection.Take.time_contiguous((2, 1000, 1), &apos;raise&apos;, &apos;longfloat&apos;)
</pre>

eric-wieser commented on numpy/core/src/multiarray/item_selection.c:
```diff
@@ -21,11 +21,171 @@
+
+
+static NPY_INLINE int
+npy_fasttake_impl(
```
Elsewhere I've seen we have an `-O3` annotation for gcc. If we don't compile at that level by default, can we use that annotation here?

seberg replied:
I had to put the attribute on the `npy_fasttake` function and not on the `_impl` one (or maybe the NPY_INLINE has to have a different order (come first). But I think it broke the inlinening first. In any case that works now. The slowdown with complex256 is still there (no idea...), but the general PutMask slowdown is much less/gone and some things got quite a bit faster.

seberg replied:
(the timings in the above comment are updated.)

seberg commented:
OK, I hope this is more or less converging. There is a slightly slowdown for certain takes, but I am not too concerned, and it could be adapted later if necessary. The complex256 slowdown is somewhat annoying, but I do not think it really matters much and may well just get magically fixed with other compilers or so.

seberg commented:
Hmmm, the azure error seems real. While float96 should not be optimized, something goes horribly wrong. Does `memmove` have alignment requirements, or is this a compiler bug with the `-O3` optimization level? Grrr....

seberg commented:
Oooops, it was a stupid typo... The issue was only that none of the tests hit the default path, except on the 32bit (actually linux) platform.

seberg commented:
From the call: It sounds like nobody is too concerned about ignoring these slots and a slight difference in speed trade-offs (and I guess if someone notices slowdowns we can optimize). Technical review awaiting of course.

mattip commented on doc/release/upcoming_changes/14942.compatibility.rst:
```diff
@@ -0,0 +1,6 @@
+Fasttake and fastputmask slots are deprecated and NULL'ed
+---------------------------------------------------------
+The fasttake and fastputmask slots are now never used internally.
+They will always set to NULL for builtin dtypes.
```
"They will always be NULL" or "They will always be set to NULL ..."

mattip commented:
In `PyArray_RegisterDataType`, maybe error out if the fields are not NULL?

seberg commented:
Sounds like a good idea, although maybe a warning (forever) that it is just ignored is just as good? Removes any possible breaking, and at some point we have to deprecate that whole function anyway...

mattip commented:
> any possible breaking

Up to you. I would argue that
- anyone taking the effort to write a function that is then ignored might appreciate getting an error rather than a warning
- it is hard to find a non-existent call to their function in the code base, leading to a hard-to-debug situation where their code is never called by NumPy but may be called by some third party package

When we remove the functions the error will go away too.

seberg committed:
DEP: Actually deprecate fasttake and fastclipmask

seberg commented on numpy/core/src/multiarray/usertypes.c:
```diff
@@ -176,6 +202,11 @@ PyArray_RegisterDataType(PyArray_Descr *descr)
         return -1;
     }
+
+    if (test_deprecated_arrfuncs_members(f) < 0) {
```
Put into its own helper in anticiapation of more functions joining. In particular `fastclip` is already deprecated at execution time, and should be added here as well now (as well as get the same documentation updates). Planning to do that once this is in.

seberg commented:
Testing this seems a bit annoying, so I only tested all the cases manually right now... I can try to think of a way to test it reasonably...

seberg commented:
I opted for warning, the alternative is to use the `clip` warning at the place where the calculation happens, but I do not think that makes sense here: chances are you are not actually doing that calculation, and the results _should never change_. You are right for someone who writes a dtype, I am thinking of the possibility of someone only using one.

As mentioned above: I think clip should be deprecated in both places (here would be enough I suppose, but I do not feel like removing the other one), I will do that as a followup.

mattip commented:
Thanks @seberg
